File: leetcode34/leetcode34.cpp
vector<int> searchRange(vector<int>& nums, int target) {
    auto lower = findBound(nums, target, true);
    auto upper = findBound(nums, target, false);
    return { lower, upper };
}

int findBound(vector<int>& nums, int target, bool isFirst) {
    int left = 0, right = nums.size() - 1, bound = -1;
    while (left <= right) {
        int mid = left + (right - left) / 2;
        if (nums[mid] == target) {
            bound = mid;
            if (isFirst) right = mid - 1;
            else left = mid + 1;
        } else if (nums[mid] < target) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }
    return bound;
}

File: leetcode34/leetcode34.py
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        def findIndex(isFirst):
            left, right = 0, len(nums) - 1
            index = -1
            while left <= right:
                mid = (left + right) // 2
                if nums[mid] == target:
                    index = mid
                    if isFirst:
                        right = mid - 1
                    else:
                        left = mid + 1
                elif nums[mid] < target:
                    left = mid + 1
                else:
                    right = mid - 1
            return index
        first = findIndex(True)
        last = findIndex(False)
        return [first, last]       
